# HELIX Revenue Prediction

## Notebook 01 — Dataset Audit

Author : Hilmi Aji

Framework : HELIX AI Shop Framework

Objective

Dataset audit before Exploratory Data Analysis (EDA)
to know the quality of data and business strategy

Output

- Dataset overview
- Data quality report
- Business data dictionary
- Initial modeling recommendation

In [ ]:
# import library
from pathlib import Path

import warnings

warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

import seaborn as sns

from IPython.display import display

plt.style.use("ggplot")

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 200)

In [ ]:
# load dataset
PROJECT_ROOT = Path("..")

DATA_PATH = PROJECT_ROOT / "data" / "raw" / "ecommerce_customer_behavior_dataset_v2.csv"

df = pd.read_csv(DATA_PATH)

print(df.shape)

df.head()

(17049, 18)


,Order_ID,Customer_ID,Date,Age,Gender,City,Product_Category,Unit_Price,Quantity,Discount_Amount,Total_Amount,Payment_Method,Device_Type,Session_Duration_Minutes,Pages_Viewed,Is_Returning_Customer,Delivery_Time_Days,Customer_Rating
0,ORD_000001-1,CUST_00001,2023-05-29,40,Male,Ankara,Books,29.18,1,0.00,29.18,Digital Wallet,Mobile,14,9,True,13,4
1,ORD_000001-2,CUST_00001,2023-10-12,40,Male,Ankara,Home & Garden,644.40,1,138.05,506.35,Credit Card,Desktop,14,8,True,6,2
2,ORD_000001-3,CUST_00001,2023-12-05,40,Male,Ankara,Sports,332.82,5,0.00,1664.10,Credit Card,Mobile,15,10,True,9,4
3,ORD_000002-1,CUST_00002,2023-05-11,33,Male,Istanbul,Food,69.30,5,71.05,275.45,Digital Wallet,Desktop,16,13,True,4,4
4,ORD_000002-2,CUST_00002,2023-06-16,33,Male,Istanbul,Beauty,178.15,3,0.00,534.45,Credit Card,Mobile,14,7,True,6,4


In [ ]:
# dataset overview
display(df.head())

display(df.tail())

display(df.sample(5, random_state=42))

,Order_ID,Customer_ID,Date,Age,Gender,City,Product_Category,Unit_Price,Quantity,Discount_Amount,Total_Amount,Payment_Method,Device_Type,Session_Duration_Minutes,Pages_Viewed,Is_Returning_Customer,Delivery_Time_Days,Customer_Rating
0,ORD_000001-1,CUST_00001,2023-05-29,40,Male,Ankara,Books,29.18,1,0.00,29.18,Digital Wallet,Mobile,14,9,True,13,4
1,ORD_000001-2,CUST_00001,2023-10-12,40,Male,Ankara,Home & Garden,644.40,1,138.05,506.35,Credit Card,Desktop,14,8,True,6,2
2,ORD_000001-3,CUST_00001,2023-12-05,40,Male,Ankara,Sports,332.82,5,0.00,1664.10,Credit Card,Mobile,15,10,True,9,4
3,ORD_000002-1,CUST_00002,2023-05-11,33,Male,Istanbul,Food,69.30,5,71.05,275.45,Digital Wallet,Desktop,16,13,True,4,4
4,ORD_000002-2,CUST_00002,2023-06-16,33,Male,Istanbul,Beauty,178.15,3,0.00,534.45,Credit Card,Mobile,14,7,True,6,4


,Order_ID,Customer_ID,Date,Age,Gender,City,Product_Category,Unit_Price,Quantity,Discount_Amount,Total_Amount,Payment_Method,Device_Type,Session_Duration_Minutes,Pages_Viewed,Is_Returning_Customer,Delivery_Time_Days,Customer_Rating
17044,ORD_004999-1,CUST_04999,2024-01-16,44,Male,Antalya,Beauty,68.65,2,0.00,137.30,Digital Wallet,Mobile,14,10,False,6,1
17045,ORD_005000-1,CUST_05000,2023-02-22,24,Female,Eskisehir,Beauty,156.12,1,0.00,156.12,Credit Card,Mobile,15,8,True,7,5
17046,ORD_005000-2,CUST_05000,2023-06-29,24,Female,Eskisehir,Home & Garden,1065.24,3,0.00,3195.72,Credit Card,Desktop,17,11,True,5,4
17047,ORD_005000-3,CUST_05000,2023-12-29,24,Female,Eskisehir,Books,47.73,2,8.80,86.66,Credit Card,Desktop,13,11,True,12,2
17048,ORD_005000-4,CUST_05000,2024-01-19,24,Female,Eskisehir,Home & Garden,1015.23,4,217.01,3843.91,Credit Card,Mobile,15,7,True,3,4


,Order_ID,Customer_ID,Date,Age,Gender,City,Product_Category,Unit_Price,Quantity,Discount_Amount,Total_Amount,Payment_Method,Device_Type,Session_Duration_Minutes,Pages_Viewed,Is_Returning_Customer,Delivery_Time_Days,Customer_Rating
450,ORD_000142-4,CUST_00142,2023-06-22,50,Male,Adana,Electronics,473.32,1,0.00,473.32,Debit Card,Mobile,15,11,True,4,3
15779,ORD_004638-3,CUST_04638,2024-02-05,20,Female,Izmir,Beauty,101.98,1,0.00,101.98,Debit Card,Desktop,22,6,True,5,3
11689,ORD_003428-6,CUST_03428,2024-02-07,51,Female,Izmir,Toys,125.66,5,46.63,581.67,Cash on Delivery,Mobile,11,7,True,1,4
5479,ORD_001580-1,CUST_01580,2024-01-25,25,Male,Izmir,Books,23.26,2,0.00,46.52,Bank Transfer,Desktop,18,7,True,3,3
1769,ORD_000549-1,CUST_00549,2023-08-05,39,Female,Adana,Sports,394.77,2,0.00,789.54,Credit Card,Mobile,15,11,True,5,4


In [4]:
# dataset dimension
rows, cols = df.shape

summary = pd.DataFrame({"Metric": ["Rows", "Columns"], "Value": [rows, cols]})

display(summary)

,Metric,Value
0,Rows,17049
1,Columns,18


In [5]:
# dataset information
df.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17049 entries, 0 to 17048
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Order_ID                  17049 non-null  object 
 1   Customer_ID               17049 non-null  object 
 2   Date                      17049 non-null  object 
 3   Age                       17049 non-null  int64  
 4   Gender                    17049 non-null  object 
 5   City                      17049 non-null  object 
 6   Product_Category          17049 non-null  object 
 7   Unit_Price                17049 non-null  float64
 8   Quantity                  17049 non-null  int64  
 9   Discount_Amount           17049 non-null  float64
 10  Total_Amount              17049 non-null  float64
 11  Payment_Method            17049 non-null  object 
 12  Device_Type               17049 non-null  object 
 13  Session_Duration_Minutes  17049 non-null  int64  
 14  Pages_

In [6]:
# memory usage
memory = df.memory_usage(deep=True).to_frame("Memory (Bytes)")

memory["Memory (MB)"] = memory["Memory (Bytes)"] / 1024 / 1024

display(memory)

print(f"Total Memory : {memory['Memory (MB)'].sum():.2f} MB")

,Memory (Bytes),Memory (MB)
Index,132,0.000126
Order_ID,1040083,0.991900
Customer_ID,1005891,0.959292
Date,1005891,0.959292
Age,136392,0.130074
Gender,921083,0.878413
City,948024,0.904106
Product_Category,953949,0.909757
Unit_Price,136392,0.130074
Quantity,136392,0.130074


Total Memory : 8.68 MB


In [7]:
# missing value audit
missing = pd.DataFrame(
    {"Missing": df.isna().sum(), "Percentage": (df.isna().mean() * 100).round(2)}
)

missing = missing.sort_values("Missing", ascending=False)

display(missing)

,Missing,Percentage
Order_ID,0,0.0
Customer_ID,0,0.0
Date,0,0.0
Age,0,0.0
Gender,0,0.0
City,0,0.0
Product_Category,0,0.0
Unit_Price,0,0.0
Quantity,0,0.0
Discount_Amount,0,0.0


In [8]:
# duplicate audit
duplicate = df.duplicated().sum()

print(f"Duplicate Rows : {duplicate}")

Duplicate Rows : 0


In [9]:
# unique value audit
unique = pd.DataFrame({"Unique": df.nunique()})

display(unique)

,Unique
Order_ID,17049
Customer_ID,5000
Date,450
Age,57
Gender,3
City,10
Product_Category,8
Unit_Price,14711
Quantity,5
Discount_Amount,5526


In [10]:
# cardinality audit
cardinality = df.nunique().sort_values(ascending=False).to_frame("Unique Values")

display(cardinality)

,Unique Values
Order_ID,17049
Total_Amount,15929
Unit_Price,14711
Discount_Amount,5526
Customer_ID,5000
Date,450
Age,57
Delivery_Time_Days,25
Session_Duration_Minutes,23
Pages_Viewed,18


In [11]:
# numerical feature
num_cols = df.select_dtypes(include=np.number).columns

display(df[num_cols].describe().T)

,count,mean,std,min,25%,50%,75%,max
Age,17049.0,34.945745,11.046855,18.00,26.00,35.00,42.00,75.00
Unit_Price,17049.0,447.901689,722.319705,5.05,73.26,174.68,494.57,7900.01
Quantity,17049.0,3.011379,1.417027,1.00,2.00,3.00,4.00,5.00
Discount_Amount,17049.0,69.788135,240.704662,0.00,0.00,0.00,32.71,6538.29
Total_Amount,17049.0,1277.438711,2358.436375,6.21,172.97,455.85,1267.75,37852.05
Session_Duration_Minutes,17049.0,14.535633,2.925524,4.00,13.00,15.00,17.00,26.00
Pages_Viewed,17049.0,9.003109,2.259954,1.00,7.00,9.00,11.00,18.00
Delivery_Time_Days,17049.0,6.503607,3.488787,1.00,4.00,6.00,8.00,25.00
Customer_Rating,17049.0,3.899408,1.128803,1.00,3.00,4.00,5.00,5.00


In [12]:
# categorical feature
cat_cols = df.select_dtypes(include="object").columns

for col in cat_cols:
    print("=" * 80)

    print(col)

    display(df[col].value_counts().to_frame())

Order_ID


,count
Order_ID,
ORD_000001-1,1
ORD_000001-2,1
ORD_000001-3,1
ORD_000002-1,1
ORD_000002-2,1
...,...
ORD_004999-1,1
ORD_005000-1,1
ORD_005000-2,1


Customer_ID


,count
Customer_ID,
CUST_02860,10
CUST_02004,10
CUST_00016,10
CUST_02572,10
CUST_00489,10
...,...
CUST_04052,1
CUST_04053,1
CUST_04059,1


Date


,count
Date,
2023-07-29,59
2023-02-14,56
2024-01-15,54
2023-05-25,53
2023-09-23,53
...,...
2023-03-04,24
2023-06-17,24
2024-02-21,24


Gender


,count
Gender,
Female,8613
Male,8176
Other,260


City


,count
City,
Istanbul,4402
Ankara,2422
Izmir,2072
Bursa,1721
Adana,1326
Antalya,1246
Gaziantep,1183
Konya,1120
Kayseri,851


Product_Category


,count
Product_Category,
Sports,2248
Beauty,2212
Books,2206
Food,2103
Toys,2090
Electronics,2074
Home & Garden,2060
Fashion,2056


Payment_Method


,count
Payment_Method,
Credit Card,6801
Debit Card,4321
Digital Wallet,3276
Bank Transfer,1763
Cash on Delivery,888


Device_Type


,count
Device_Type,
Mobile,9543
Desktop,5845
Tablet,1661


In [13]:
# datetime
df["Date"] = pd.to_datetime(df["Date"])

print(df["Date"].min())

print(df["Date"].max())

2023-01-01 00:00:00
2024-03-25 00:00:00


In [14]:
# target audit
target = "Total_Amount"

display(df[target].describe())

print("Skewness:", df[target].skew())

print("Kurtosis:", df[target].kurtosis())

count    17049.000000
mean      1277.438711
std       2358.436375
min          6.210000
25%        172.970000
50%        455.850000
75%       1267.750000
max      37852.050000
Name: Total_Amount, dtype: float64

Skewness: 4.584841564259209
Kurtosis: 30.380082777089246


In [15]:
# initial leakage detection
check = df["Unit_Price"] * df["Quantity"] - df["Discount_Amount"]

difference = (check - df["Total_Amount"]).abs().max()

print(f"Maximum Difference : {difference:.10f}")

Maximum Difference : 0.0000000000


# Business Data Dictionary

| Feature | Business Category | Description | Role | Initial Decision |
|----------|-------------------|-------------|------|------------------|
| Order_ID | Transaction | Unique identifier for each order | Identifier | Drop |
| Customer_ID | Customer | Unique identifier for each customer | Identifier | Drop (Version 1) |
| Date | Time | Transaction date | Feature | Feature Engineering |
| Age | Customer | Customer age | Numerical | Keep |
| Gender | Customer | Customer gender | Categorical | Keep |
| City | Geography | Customer city | Categorical | Keep |
| Product_Category | Product | Purchased product category | Categorical | Keep |
| Unit_Price | Transaction | Product unit price | Numerical | Investigate Target Leakage |
| Quantity | Transaction | Number of purchased items | Numerical | Investigate Target Leakage |
| Discount_Amount | Promotion | Discount applied to the order | Numerical | Investigate Target Leakage |
| Total_Amount | Revenue | Total transaction amount | Target | Prediction Target |
| Payment_Method | Transaction | Payment method used | Categorical | Keep |
| Device_Type | Customer Behavior | Device used during purchase | Categorical | Keep |
| Session_Duration_Minutes | Customer Behavior | Website browsing duration | Numerical | Keep |
| Pages_Viewed | Customer Behavior | Number of visited pages | Numerical | Keep |
| Is_Returning_Customer | Customer | Customer loyalty status | Boolean | Keep |
| Delivery_Time_Days | Logistics | Delivery duration | Numerical | Keep |
| Customer_Rating | Satisfaction | Customer satisfaction score | Numerical | Keep |

---

## Initial Feature Categorization

### Identifier Features

- Order_ID
- Customer_ID

These features uniquely identify records but are not expected to contribute to prediction performance. They will be excluded from the first modeling iteration.

---

### Customer Features

- Age
- Gender
- Is_Returning_Customer

These variables describe customer demographics and loyalty behavior.

---

### Behavioral Features

- Device_Type
- Session_Duration_Minutes
- Pages_Viewed

These variables represent customer interaction before completing a purchase.

---

### Transaction Features

- Unit_Price
- Quantity
- Discount_Amount
- Payment_Method

These variables describe the transaction itself and require further investigation for potential target leakage.

---

### Product Features

- Product_Category

Represents the purchased product category.

---

### Geographic Features

- City

Represents customer location.

---

### Logistics Features

- Delivery_Time_Days

Represents delivery performance.

---

### Satisfaction Features

- Customer_Rating

Represents customer satisfaction after purchase.

---

### Time Features

- Date

Will be transformed into multiple temporal features during Feature Engineering.

# 19. Initial Findings

## Dataset Quality Assessment

The dataset is considered to be in excellent condition for exploratory analysis.

### Positive Findings

- The dataset contains **17,049 observations** and **18 variables**, providing sufficient data for machine learning experiments.
- No missing values were detected across all features.
- No duplicated records were identified.
- All columns have appropriate data types and can be processed without major structural modifications.
- The transaction date is complete and suitable for temporal feature engineering.

---

## Target Variable Assessment

The prediction target is:

**Total_Amount**

Initial statistical analysis indicates that:

- The distribution is highly right-skewed.
- The mean is considerably higher than the median, suggesting the presence of high-value transactions.
- Outlier analysis will be performed in the next notebook to determine whether transformation or robust modeling techniques are required.

---

## Potential Target Leakage

A preliminary investigation suggests that:

The computed values are nearly identical to the target, indicating that these three variables may directly determine the prediction target.

If confirmed, using these variables in the predictive model would produce artificially high performance while providing limited business value.

Therefore, two modeling scenarios are proposed:

### Scenario A — Business Revenue Prediction

Predict revenue using customer characteristics and behavioral features only.

Candidate features include:

- Age
- Gender
- City
- Product_Category
- Payment_Method
- Device_Type
- Session_Duration_Minutes
- Pages_Viewed
- Is_Returning_Customer
- Delivery_Time_Days
- Customer_Rating
- Engineered datetime features

This scenario reflects a realistic business prediction problem.

---

### Scenario B — Transaction Amount Estimation

Include:

- Unit_Price
- Quantity
- Discount_Amount

This scenario estimates transaction totals once purchase information is already available.

Although prediction accuracy is expected to be substantially higher, the resulting model is less valuable for early-stage business decision making.

---

## Recommended Next Step

Proceed with a comprehensive Exploratory Data Analysis (EDA), including:

1. Univariate analysis
2. Categorical feature analysis
3. Target distribution analysis
4. Correlation analysis
5. Outlier detection
6. Datetime analysis
7. Target leakage verification

The results of these analyses will guide data cleaning, feature engineering, and model development in subsequent notebooks.